### PR TITLE
Fixes and updates in preparation for ROOT 6.12 release

### PR DESCRIPTION
**Pull request by amadio**
Just testing for now, please do not merge as there are LLVM patches in the mix.

### CODE REVIEW

**Comment by amadio** [comment]
Ok, Jenkins doesn't like the changes that disable the builtin versions by default, since it enables `fail-on-missing` for the builds... I will do that in a separate PR later, as it will require to change the configuration in Jenkins too.

**Comment by amadio** [comment]
I don't think the failure on Windows is caused by this pull request.

**Comment by pcanal** [comment]
>  Remove unimplemented C++17 custom new/delete operators 

Are you sure this is the right direction?  As far as I know/recall those will actually be needed (to be implemented) to be complete/correct.  (Not implementing them may lead to problem similar to those that lead to 0fc4f521b64c927977460253dc1e622eda431021 ... compilation works but behavior was incorrect)

**Comment by amadio** [comment]
> Are you sure this is the right direction?

Yes and no. I first got ROOT to compile and work. That commit just allows to compile. Before the release we need to revert this and add back proper implementations of the custom `new/delete` operators so that we can track the memory. Having them there with the wrong signature and returning `nullptr` is actually worse than not having them, though. Whoever did that should have spent the extra hour or so of work to properly implement them instead.

**Comment by amadio** [comment]
Also, this is a total shame:
```
root $ git grep 'not implemented' | wc -l
672
```

**Comment by pcanal** [comment]
> Having them there with the wrong signature and returning nullptr is actually worse than not having them, though. 

Not having them leads to hard to track memory allocation/deletion mismatch in some cases.  The best is indeed to have an implementation but a clear error (maybe throwing would be even better) is 'better' :)

**Comment by pcanal** [comment]
> root $ git grep 'not implemented' | wc -l

Most of these are a C++98 implementation of '=delete' to prevent copy of object. (i.e. hide/delete the copy constructors).

**Comment by amadio** [comment]
>> root $ git grep 'not implemented' | wc -l

> Most of these are a C++98 implementation of '=delete' to prevent copy of object. (i.e. hide/delete the copy constructors).

At least half of them are not comments, and at least 150 or so occurrences are for actual non-implemented stuff (i.e. inside a `Fatal` or `Error` call or something similar). I checked that afterwards.

**Comment by amadio** [comment]
I've now implemented the operator new and delete instead of just deleting them. If the build passes, I will merge this today so that others can try to build with C++17 enabled with Clang-5.0 to confirm that these changes are enough. Also, we need to decide if the commits on LLVM should be applied on their side first ot not. @vgvassilev @Axel-Naumann Please let me know.

File: cmake/modules/RootBuildOptions.cmake [hunk 1]
```diff
@@ -99,7 +99,7 @@ ROOT_BUILD_OPTION(cocoa OFF "Use native Cocoa/Quartz graphics backend (MacOS X o
 ROOT_BUILD_OPTION(cuda OFF "Use CUDA if it is found in the system")
 ROOT_BUILD_OPTION(cxx11 ON "Build using C++11 compatible mode, requires gcc > 4.7.x or clang")
 ROOT_BUILD_OPTION(cxx14 OFF "Build using C++14 compatible mode, requires gcc > 4.9.x or clang")
-#ROOT_BUILD_OPTION(cxx17 OFF "Build using C++17 compatible mode, requires gcc >= 7.1.0 or clang")
+ROOT_BUILD_OPTION(cxx17 OFF "Build using C++17 compatible mode, requires gcc >= 7.2.0 or clang")
```
Comment (amadio) [inline cmake/modules/RootBuildOptions.cmake]: I should actually make this `GCC >= 7.3.0 or Clang >= 5.0`, I guess.

File: core/meta/src/TGenericClassInfo.cxx [hunk 2]
```diff
@@ -27,31 +27,41 @@
 namespace ROOT {
 namespace Internal {
    std::string TTypeNameExtractionBase::GetImpl(const char* derived_funcname) {
```
Comment (amadio) [inline core/meta/src/TGenericClassInfo.cxx]: Note, this function still doesn't work on Windows, and now my changes make ROOT error out if type deduction fails, so Windows builds are going to fail. In the long term, I would prefer to get and use the mangled name as is, without parsing compiler output. Parsing like done here is quite unreliable, as this function itself proves. @bellenot , could you please let us know what Visual Studio prints so we can fix it?

Comment (amadio) [reply core/meta/src/TGenericClassInfo.cxx]: Nevermind, it already failed with
```
Fatal in <TClassTable::Add()>: Failed to deduce type for 'class ROOT::Internal::TCheckHashRecurveRemoveConsistency'
```
So I will add code to gobble up `"class "` in the beggining of the type string and see if that works.

Comment (amadio) [reply core/meta/src/TGenericClassInfo.cxx]: Hi @bellenot. Thanks. ROOT would just add the types that failed with an empty string as the key before. With ICC that started to generate lots of warnings and test failures, so I turned it into a fatal error.

Comment (pcanal) [reply core/meta/src/TGenericClassInfo.cxx]: > In the long term, I would prefer to get and use the mangled name as is, without parsing compiler output. 

Agreed.  It is 'just' a matter of using TClassEdit::DemangleTypeIdName (on the right type :) ), so this might be straightforward enough for this PR.

File: core/newdelete/src/NewDelete.cxx [hunk 3]
```diff
@@ -257,6 +241,24 @@ void *operator new(size_t size, void *vp)
 }
 #endif
 
+#if __cplusplus >= 201700L
+void *operator new(size_t size, std::align_val_t al)
+{
+   void *ptr = nullptr;
+   if (posix_memalign(&ptr, static_cast<size_t>(al), size) != 0)
```
Comment (pcanal) [inline core/newdelete/src/NewDelete.cxx]: This makes the custom operator new with and without alignment incompatible/semantically-different.  

Note that one of the feature of the custom operator new is that when gMmallocDesc is set it will use (indirectly) mmalloc (allocation of shared memory) rather than calloc/malloc.

And since this overload will/might call implicitly depending on the circunstance, things that used to be allocated in the 'shared memory section' will (silently) no longer be.

File: core/meta/src/TGenericClassInfo.cxx (hunk 2 above)
Comment (amadio) [reply core/meta/src/TGenericClassInfo.cxx]: I will try, but why are we doing it the way it is now if it was just a matter of using `TClassEdit::DemangleTypeIdName`? I do not know what is needed as the key, so I did not try to change it. That is, I don't know if changing it from mangled to demandled has any effect, if it just needs to be something unique, etc. I need you and @Axel-Naumann to help me here.

Comment (pcanal) [reply core/meta/src/TGenericClassInfo.cxx]: >  I do not know what is needed as the key

We need the demangled fully qualified name of the class (T) to be returned by TTypeNameExtraction<T>::Get().  A priori this could be accomplish by passing the typeid(T) to GetImp and have GetImp call TClassEdit::DemangleTypeIdName.

> That is, I don't know if changing it from mangled to demandled has any effect

The name currently returned is demangled.

> why are we doing it the way it is now

That seems to have been lost in time ... It is plausible that at the time the author concluded that TClassEdit could not be used from Rtypes.h (possibly due to the way the files were organized at the time).   We probably should still not #include TClassEdit.h in Rtypes.h but we can have TGenericClassInfo.cxx do it safely.

File: core/newdelete/src/NewDelete.cxx (hunk 3 above)
Comment (amadio) [reply core/newdelete/src/NewDelete.cxx]: This is not yet used anywhere. The comments you made, although they are correct, do not give me the direction of what a correct solution would be, since the implementation of the custom placement new seems incorrect. What I care about at the moment is that experiments are waiting for a version of ROOT that compiles with C++17 support, and this (something they don't need) is blocking them. I do not think that ROOT should be replacing the allocation operators in the global namespace for all user code, and therefore my first attempt was to simply remove the "implementations" that you put in, since they don't compile. If ROOT needs to have custom operators, we can add them later into the ROOT namespace and update the code that needs them (which for now is still no code at all).

File: core/meta/src/TGenericClassInfo.cxx (hunk 2 above)
Comment (amadio) [reply core/meta/src/TGenericClassInfo.cxx]: What about this? Is it correct behavior?
```
   ------------------------------------------------------------
  | Welcome to ROOT 6.11/02                http://root.cern.ch |
  |                               (c) 1995-2017, The ROOT Team |
  | Built for linuxx8664gcc                                    |
  | From tag v6-11-02, 6 October 2017                          |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q' |
   ------------------------------------------------------------

root [0] int status;
root [1] cout << TClassEdit::DemangleTypeIdName(typeid(Double32_t), status) << endl;
double
```

Comment (amadio) [reply core/meta/src/TGenericClassInfo.cxx]: Answer = yes?
```
root [5] cout << ROOT::Internal::TTypeNameExtraction<Double32_t>::Get() << endl;
double
```

File: core/newdelete/src/NewDelete.cxx (hunk 3 above)
Comment (pcanal) [reply core/newdelete/src/NewDelete.cxx]: >  do not think that ROOT should be replacing the allocation operators in the global namespace

But it already does, optionally (if you load -lNew).

> and therefore my first attempt was to simply remove the "implementations"

The point of those (intentionally broken since at the time there was no compiler to test them) implementation was as a remainder to do the right thing.   

We can do either of two things

1) overload operator new with alignment and make it fail (Fatal or exception)
2) implement properly.

> since the implementation of the custom placement new seems incorrect.

The implementation is **correct**  but the behavior is *of course* different from regular operator new.  The custom operator new without placement will allocate memory and need to *optionally* allocate this memory in the shared memory section (mmalloc).

> update the code that needs them (which for now is still no code at all).

Humm ... there exit code that really on the custom operator new (anything that uses TMapFile for example) ... It is clear that we do not *have to* support this for operator new with alignment but it needs to fail nicely (not silently) so option 1) is plausible.  Not implementing them at all is not an option because in that case **when the user** load libNew and (intentionally or not) updates its classes to now be set in a way that the compiler will use the operator new with alignment (whereas it was not doing so before), the program will now silently fail (allocating the object in regular memory rather than the shared memory as intended by the user).


So in conclusion we need the operator new without placement and with alignment to ***either***

a) throw an exception if ROOT::Internal::gMmallocDesc is true. [in which case the code you have will not do what the user expects/want/need]

b) use an implementation similar to the custom operator new (i.e. call mmalloc (or equivalent) when ROOT::Internal::gMmallocDesc is true)

Cheers,
Philippe.

File: core/meta/src/TGenericClassInfo.cxx (hunk 2 above)
Comment (pcanal) [reply core/meta/src/TGenericClassInfo.cxx]: > What about this? Is it correct behavior?

Unfortunately it is not ... However you point out it is no worse that the current situation.

Either way this means that ClassDefInline is currently *not* supported for class template whose argument might be Double32_t

**Comment by amadio** [comment]
Partially merged by hand. Closing.
